d2/src/main/java/com/linkedin/d2/balancer/util/WarmUpLoadBalancer.java

```diff
@@ -114,91 +115,117 @@ public WarmUpLoadBalancer(LoadBalancerWithFacilities balancer, WarmUpService ser
-        // data of service/cluster/uri (in the WarmUpTask below), so that when the actual data is received, they can
-        // be reported to dual read monitoring under dual read mode.
-        DualReadModeProvider dualReadModeProvider = _dualReadStateManager.getDualReadModeProvider();
-        _servicesToWarmUp = serviceNames.stream().filter(serviceName -> {
-          DualReadModeProvider.DualReadMode dualReadMode = dualReadModeProvider.getDualReadMode(serviceName);
-          _dualReadStateManager.updateService(serviceName, dualReadMode);
-
-          boolean res = isModeToWarmUp(dualReadMode, _isIndis);
-          if (!res)
-          {
-            LOG.info("{} skipping service: {} based on its dual read mode: {}",
-                _printName, serviceName, dualReadMode);
-          }
-          return res;
-        }).collect(Collectors.toList());
-
-        _servicesToWarmUp.forEach(serviceName -> {
-          // To warm up the cluster dual read mode, we need to fetch the service data to know its belonging cluster.
-          LOG.info("{} fetching service data for service: {}", _printName, serviceName);
-
-          // NOTE: This call blocks!
-          getLoadBalancedServiceProperties(serviceName, new Callback<ServiceProperties>() {
-            @Override
-            public void onError(Throwable e) {
-              LOG.warn("{} failed to warm up dual read mode for service: {}", _printName, serviceName, e);
-            }
+        _servicesToWarmUp = serviceNames;
 
-            @Override
-            public void onSuccess(ServiceProperties result) {
-              _dualReadStateManager.updateCluster(result.getClusterName(),
-                  _dualReadStateManager.getServiceDualReadMode(result.getServiceName()));
+        if (_dualReadStateManager != null)
+        {
+          // warm up dual read mode for the service and its belonging cluster. This is needed BEFORE fetching the actual
+          // data of service/cluster/uri (in the WarmUpTask below), so that when the actual data is received, they can
+          // be reported to dual read monitoring under dual read mode.
+          DualReadModeProvider dualReadModeProvider = _dualReadStateManager.getDualReadModeProvider();
+          _servicesToWarmUp = serviceNames.stream().filter(serviceName -> {
+            DualReadModeProvider.DualReadMode dualReadMode = dualReadModeProvider.getDualReadMode(serviceName);
+            _dualReadStateManager.updateService(serviceName, dualReadMode);
+
+            boolean res = isModeToWarmUp(dualReadMode, _isIndis);
+            if (!res)
+            {
+              LOG.info("{} skipping service: {} based on its dual read mode: {}",
+                  _printName, serviceName, dualReadMode);
+            }
+            return res;
+          }).collect(Collectors.toList());
+
+          _servicesToWarmUp.forEach(serviceName -> {
+            // check timeout before continue
+            if (!hasTimedOut.get()
```

Comment:
The variable doesn't need to be atomic since it's only accessed within the executor and is actually local. This operation (read and then update) is in itself non-atomic and thus thread-unsafe even if an atomic was required.